pyroscope/session.go

```diff
@@ -186,6 +221,96 @@ func (ps *Session) uploadData(startTime, endTime time.Time) {
 		ps.cpuBuf.Reset()
 	}
 
+	if ps.isGoroutinesEnabled() {
+		p := pprof.Lookup("goroutine")
+		if p != nil {
+			p.WriteTo(ps.goroutinesBuf, 0)
+			ps.upstream.Upload(&upstream.UploadJob{
+				Name:            ps.appName,
+				StartTime:       startTime,
+				EndTime:         endTime,
+				SpyName:         "gospy",
+				Units:           "goroutines",
+				AggregationType: "average",
+				Format:          upstream.FormatPprof,
+				Profile:         copyBuf(ps.goroutinesBuf.Bytes()),
+				SampleTypeConfig: map[string]*upstream.SampleType{
+					"goroutine": {
+						DisplayName: "goroutines",
+						Units:       "goroutines",
+						Aggregation: "average",
+					},
+				},
+			})
+			ps.goroutinesBuf.Reset()
+		}
+	}
+
+	if ps.isBlockEnabled() {
+		p := pprof.Lookup("block")
+		if p != nil {
+			p.WriteTo(ps.blockBuf, 0)
+			curBlockBuf := copyBuf(ps.blockBuf.Bytes())
+			ps.blockBuf.Reset()
+			if ps.blockPrevBytes != nil {
+				ps.upstream.Upload(&upstream.UploadJob{
+					Name:        ps.appName,
+					StartTime:   startTime,
+					EndTime:     endTime,
+					SpyName:     "gospy",
+					Format:      upstream.FormatPprof,
+					Profile:     curBlockBuf,
+					PrevProfile: ps.blockPrevBytes,
+					SampleTypeConfig: map[string]*upstream.SampleType{
+						"contentions": {
+							DisplayName: "block_count",
+							Units:       "lock_samples",
+							Cumulative:  true,
+						},
+						"delay": {
+							DisplayName: "block_duration",
+							Units:       "lock_nanoseconds",
+							Cumulative:  true,
+						},
+					},
+				})
+			}
+			ps.blockPrevBytes = curBlockBuf
+		}
+	}
+	if ps.isMutexEnabled() {
+		p := pprof.Lookup("mutex")
+		if p != nil {
+			p.WriteTo(ps.mutexBuf, 0)
+			curMutexBuf := copyBuf(ps.mutexBuf.Bytes())
+			ps.mutexBuf.Reset()
+			if ps.mutexPrevBytes != nil {
+				ps.upstream.Upload(&upstream.UploadJob{
+					Name:        ps.appName,
+					StartTime:   startTime,
+					EndTime:     endTime,
+					SpyName:     "gospy",
+					Format:      upstream.FormatPprof,
+					Profile:     curMutexBuf,
+					PrevProfile: ps.mutexPrevBytes,
+					SampleTypeConfig: map[string]*upstream.SampleType{
+						"contentions": {
+							DisplayName: "mutex_count",
+							Units:       "lock_samples",
+							Cumulative:  true,
+						},
+						"delay": {
+							DisplayName: "mutex_duration",
+							Units:       "lock_nanoseconds",
+							Cumulative:  true,
+						},
+					},
```

Comment:
I'm slightly concerned with the fact that `contentions` and `delay` sample types can't be enabled separately – for example, if a user specifies `ProfileMutexCount` they might not expect samples of `ProfileMutexDuration` to be collected. We have a similar problem with memory profile types: https://github.com/pyroscope-io/client/issues/13.